aesd-char-driver/main.c

```diff
@@ -47,24 +34,170 @@ int aesd_release(struct inode *inode, struct file *filp)
 ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
                 loff_t *f_pos)
 {
-    ssize_t retval = 0;
+    size_t num_bytes_in_buffer = 0;
+    size_t new_fpos = *f_pos;
+
+    struct aesd_dev *dev = filp->private_data;
+    struct aesd_buffer_entry *entry;
+
+    size_t spot_in_node = 0;
+    size_t avail_in_node_rem = 0;
+    size_t final_copy_num = 0;
+    size_t num_actual_copied = 0;
+
+    int how_many_nodes;
+    int idx;
+
     PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
-    /**
-     * TODO: handle read
-     */
-    return retval;
+    mutex_lock(dev->lock);
+    
+    if(dev->buf->full) how_many_nodes = AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+    else how_many_nodes = (dev->buf->in_offs - dev->buf->out_offs);
+
+    for(int i = 0; i < how_many_nodes; i++){
+
+        idx = (dev->buf->out_offs + i) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+        num_bytes_in_buffer += dev->buf->entry[idx].size;
```

Comment:
@almu6129 it might be better to add a function to your circular buffer implementation to return the total number of bytes to decouple from your driver logic.